### DISCUSSION TITLE
Just another one  CS0165: Use of unassigned variable

### DISCUSSION THREAD

**Discussion by gurustron** [dotnet/csharplang#7291] [Q&A] [answered]
Was not able to find exactly the same one. Consider the following method:

        bool aBool = SomePredicate();
        if (!aBool && !OutFunction(out int value))
        {
            return;
        }
        int result = -1; //other code here...
        if (!aBool && SomeCondition(result, value)) // CS0165: Use of unassigned variable 'value'.
        {
        }

        bool SomePredicate() => false;
        bool OutFunction(out int i) => false;
        bool SomeCondition(int i, int j) => false;

While due to the same check `!aBool` and short-circuiting at point of `SomeCondition(result, value)` logically `value` is definitely assigned the compiler fails to determine that.

[Sharplab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTARgPasA26AhgELtcAvOgDKrALYBTAAoAnScACWAY24AXSXQDcjJkUUAzdNQCEfAegBkl9CYDyAVzVQArGsWsAdtVZP0izzV0ADduDgdJWlpdPQIkPQTMAHYdeMSAXximAKD5AGcHDiDhAFo4LXQAekrWNQALSVl0ZVZgSXQG+QA6HqziQ2Mzfk4rGzEpbC8ldy9qfMK1ABoQsIioqsqcEQAGOAA2FxB0AFU89tYjB09uPLzFAHNPBRXZRW5mDnaAclDwyS+un0iHEgehMkhQWwRuMZPIlKoNHR0IIAHzoAxhM6pRLoKFcRzONweby+II5fz0VHoeqyVgAd3QTwZAFEAB7KSQABxm3loqUhFhhk0802J1HJimW5IAVpS0RiOFiYuD0kA===)

**Comment by bernd5** [comment]
if `aBool` is true `OutFunction` is not executed - value is not assigned.

In the past there were already some improvements of the [Definite Assignment Analysis](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/improved-definite-assignment).

But I think it is not only hard for the compiler to see this - for humans it is hard, too.

I would refactor the code to:
```csharp
using System;
public class C {
    public void M() {
        if (!SomePredicate())
        {
            if (!OutFunction(out int value))
            {
                return;
            }
            int result = -1; //other code here...
            if (SomeCondition(result, value)) // CS0165: Use of unassigned variable 'value'.
            {

            }
        }
        
        bool SomePredicate() => false;
        bool OutFunction(out int i) => throw new Exception();

        bool SomeCondition(int i, int j) => false;
    }
}
```

**Reply by gurustron** [reply]
@bernd5 that is what I wrote in this [SO answer](https://stackoverflow.com/a/76514584/2501279) =) I was wondering if there are some additional insights here (hence like some specs or links where compiler developers explain why it expected here).

**Reply by bernd5** [reply]
See the rules: https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#94-definite-assignment

**Reply by bernd5** [reply]
See especially: https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#94426--expressions

**Reply by RikkiGibson** [reply, marked as answer]
https://github.com/dotnet/roslyn/issues/36927#issuecomment-508595947

**Reply by gurustron** [reply]
@RikkiGibson and @bernd5 thank you!
